communication-refactored/src/interface.rs

```diff
@@ -0,0 +1,559 @@
+        let dns_transport = TokioDnsConfig::system(TokioTcpConfig::new())?;
+        let transport = dns_transport.clone().or_transport(WsConfig::new(dns_transport));
+        Ok(self.build_with_transport(transport).await)
+    }
+
+    /// Create a new [`ShCommunication`] instance with an underlying [`Swarm`][libp2p::Swarm] that uses the provided
+    /// transport.
+    ///
+    /// The transport is upgraded with:
+    /// - [Relay protocol][<https://docs.libp2p.io/concepts/circuit-relay/>] (requires *feature = "relay"*)
+    /// - Authentication and encryption with the Noise-Protocol, using the XX-handshake
+    /// - Yamux substream multiplexing
+    ///
+    /// The method spawns a new task that handles all interaction with the Swarm.
+    /// The task runs until [`ShCommunication`] is dropped, [`ShCommunication`] provides an interface to perform
+    /// operations on the swarm-task. **Note:** The task is spawned using [`tokio::spawn`], hence this method has to
+    /// be called in the context of a tokio.rs runtime.
+    pub async fn build_with_transport<Tp>(self, transport: Tp) -> ShCommunication<Rq, Rs, TRq>
+    where
+        Tp: Transport + Sized + Clone + Send + Sync + 'static,
+        Tp::Output: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+        Tp::Dial: Send + 'static,
+        Tp::Listener: Send + 'static,
+        Tp::ListenerUpgrade: Send + 'static,
+        Tp::Error: Send + Sync,
+    {
+        // Use the configured keypair or create a new one.
+        let (noise_keypair, peer_id) = self.ident.unwrap_or_else(|| {
+            let keypair = Keypair::generate_ed25519();
+            let noise_keypair = NoiseKeypair::<X25519Spec>::new().into_authentic(&keypair).unwrap();
+            let peer_id = keypair.public().into_peer_id();
+            (noise_keypair, peer_id)
+        });
+        #[cfg(feature = "relay")]
+        let (transport, relay_behaviour) = new_transport_and_behaviour(RelayConfig::default(), transport);
+        let transport = transport
+            .upgrade(upgrade::Version::V1)
+            .authenticate(NoiseConfig::xx(noise_keypair).into_authenticated())
+            .multiplex(YamuxConfig::default())
+            .boxed();
+        #[cfg(feature = "mdns")]
+        let mdns = Mdns::new(MdnsConfig::default())
+            .await
+            .expect("Failed to create mdns behaviour.");
+        let behaviour = NetBehaviour::new(
+            self.behaviour_config,
+            #[cfg(feature = "mdns")]
+            mdns,
+            #[cfg(feature = "relay")]
+            relay_behaviour,
+            self.firewall_channel,
+        );
+        let mut swarm_builder = SwarmBuilder::new(transport, behaviour, peer_id).executor(Box::new(|fut| {
+            tokio::spawn(fut);
+        }));
+        if let Some(limit) = self.connections_limit {
+            swarm_builder = swarm_builder.connection_limits(limit);
+        }
+        let swarm = swarm_builder.build();
```

Comment:
```suggestion
        let swarm = swarm_builder.substream_upgrade_protocol_override(Version::V1Lazy).build();
```

Consider using `V1Lazy` instead of `V1` to save on a whole network roundtrip during stream negotiation, e.g. when doing a request.

See https://docs.rs/multistream-select/0.10.2/multistream_select/enum.Version.html#variant.V1Lazy for details.

Reply:
Thanks for the advice. I do have to admit though that the docs on `V1Lazy` are a bit difficult to understand:
> This strategy is only applicable [...] if the dialer supports just a single application protocol

What does `a single application protocol` mean? That only one version of a protocol is supported?

> Nevertheless, there is a pitfall that is rarely encountered: When nesting multiple protocol negotiations, the listener should either be known to support all of the dialer’s optimistically chosen protocols or there is must be no intermediate protocol without a payload and none of the protocol payloads must have the potential for being mistaken for a multistream-select protocol message

In what cases are protocol negotiations nested? And are there any protocols in rust-libp2p without a payload?

Reply:
> > This strategy is only applicable [...] if the dialer supports just a single application protocol
> 
> What does `a single application protocol` mean? That only one version of a protocol is supported?

Protocols on streams are negotiated using [Multistream Select](https://github.com/multiformats/multistream-select). The process is quite simple. The dialer proposes a protocol, the listener either accepts the protocol by echoing it, or rejects it by sending `na` (not available).

Say the dialer opens a new stream and only has a single protocol that it is willing to speak on that stream. E.g. not `/my-protocol/1` and `/my-protocol/2`, but only a single one, e.g. `/my-protocol/2`. What the dialer can do is send the `/my-protocol/2` and instead of waiting for the echo, or the `na`, directly send application data, not having to wait for 1 round trip.

This only works, when one is willing to speak a single protocol on that stream only. When willing to speak either one of many protocols, `V1Lazy` defaults to the `V1` behavior.


> 
> > Nevertheless, there is a pitfall that is rarely encountered: When nesting multiple protocol negotiations, the listener should either be known to support all of the dialer’s optimistically chosen protocols or there is must be no intermediate protocol without a payload and none of the protocol payloads must have the potential for being mistaken for a multistream-select protocol message
> 
> In what cases are protocol negotiations nested? And are there any protocols in rust-libp2p without a payload?

Negotiations are nested e.g. when first negotiating a security protocol, then, nested, a multiplexer and then, nested a stream protocol. `substream_upgrade_protocol_override(Version::V1Lazy)` only sets the `V1Lazy` option for the last one, not followed by any other nesting, thus safe.

I am not aware of any **intermediate** protocols without a payload.

---

This is just a suggestion. After all, playing it safe by using `V1` is just fine. You can still later on enable `V1Lazy` without having to coordinate a roll out.

Reply:
Thank you for the detailed explanation, that makes sense. Using `V1Lazy` sounds reasonable.

Reply:
For some reason, `test test_firewall::firewall_ask` does not terminate when using `V1Lazy`, and if disable the feature `tcp-transport` (which results in the use of a plain `TokioTcpConfig` as transport), `test_firewall::firewall_permissions` does not terminate as well. All other tests (dialing, sending requests) pass.

Since I don't really know what the reason for this behaviour could be, I'll use `V1` for now. 
I might investigate this further in the future, but right now it is not a priority.

Reply:
>  I might investigate this further in the future, but right now it is not a priority.

In case you get around to it, I would be curious what the cause is.